### DISCUSSION TITLE
On Unicode builds becoming the norm for Windows

### DISCUSSION THREAD

**Discussion by God-damnit-all** [curl/curl#15332] [General]
Having been using compiled Unicode builds for Windows for a long time now and having never run into any sort of complication (including libcurl use), I have to wonder: Why is this not already built by default by the curl project's own Windows distribution?

Is it because of older versions of Windows? Perhaps because of more uncommon protocols? I'd really like to know what sort of complications are standing in the way of this.

With scripts of mine that make use of curl, I usually have them download a Unicode-enabled binary just so I can be sure that filenames which include Unicode characters save as as the content disposition intends them to.

Considering how long it's been since the Unicode build option for Windows was first implemented, one has to wonder if there'll ever come a point that it becomes the norm; which, realistically, would only happen if the curl project starts distributing said builds. I really hope to see something like that happen sometime within the next few years.

**Comment by jay** [comment]
> Is it because of older versions of Windows? Perhaps because of more uncommon protocols? I'd really like to know what sort of complications are standing in the way of this.

curl's Windows Unicode support is not complete and has some bugs which is why the official build does not use it. For Unicode builds we have some basic support by opening filenames in wide characters with UTF-16 encoding. But we weren't going to do the ton of work and maintenance needed for curl to convert to wide characters everywhere, that would be like wrapping the entire CRT and Windows API. Windows now has UTF-8 support but requires a project manifest to enable it. However I don't want to enable it for non-Unicode builds. In other words I think if you build for Unicode then you should get a binary with that manifest and if not then you shouldn't. There was some disagreement and we shelved it, you can read the entire thing at #11482.

**Reply by God-damnit-all** [reply]
I appreciate the background information. Is the current state of affairs regarding this going to be touched on any time soon?

**Reply by vszakats** [reply]
I still think that the UTF-8 method (combined with non-Unicode build) would be the feasible way to go. But even that raises a list of non-trivial Windows API questions (some here: https://nullprogram.com/blog/2021/12/30/) that need to be cleared and/or worked around to make it into something ready-ish for release.

According to the article and the issue https://github.com/microsoft/terminal/issues/4551#issuecomment-585487802 linked from it, UTF-8 console input is still not implemented in Windows. That is from 2020 and was closed with the comment that it would be a "fairly big chunk of work". Backlinks from 2023 still saying it's a problem.

The article mentions `SetConsoleOutputCP()` as a requirement to make this work. It's been a long time ago, but this API caused console crashes in a different project and had to be [reverted](https://github.com/vszakats/hb/commit/4e07bf59407fd335255870609ad5e3a462559e51). This Python issue might be related: https://github.com/python/cpython/issues/45943. Perhaps it's safe to use now, but we also aim to support Windows back to XP.

(It doesn't help that testing most of this requires a native Windows 10 env.)

**Reply by vszakats** [reply]
Short updates:
- dropping Windows XP support soon may improve our options a little bit (or maybe not):
  Ref: 73d52567a0e14593cd92b4968063ae5b976f1888 #18016
- busybox-w32 went with the UTF-8 route. Here are their notes that I found useful and likely applicable to curl too:
  https://frippery.org/busybox/release-notes/FRP-5181.html
  Downloadable binary: https://frippery.org/files/busybox/busybox64u.exe
